### DISCUSSION TITLE
Background Threads as Daemon?

### DISCUSSION THREAD

**Discussion by thgau** [tkowalcz/tjahzi#75] [Q&A] [answered]
The two threads (tjahzi-worker and ReadingLogBufferAndSendingHttp) are not marked as DaemonThreads. Is there a reason for this?

**Comment by tkowalcz** [comment, marked as answer]
Default thread factory from Netty does not use daemon threads, but looks like [Netty 5.0 does](https://github.com/netty/netty/issues/2832). 

`ReadingLogBufferAndSendingHttp` thread is part of Agrona agent "framework" that also does not set threads to daemon by default. If Tjahzi is shut down correctly it should also close all its threads cleanly. Did you see any problems with that?

That being said I'm ok with changing those threads to be daemons.

**Reply by thgau** [reply]
Actually I am not sure what would be the right behavior. I was just wondering that a simple test program did not terminate after the main method ended. If the logging needs more time and should be shutdown explicitly to flush all logs, then I can live with that too.

**Reply by thgau** [reply]
You mentioned the shut down, I assume `org.apache.logging.log4j.LogManager.shutdown()`
The shutdown seems not to wait for the tjahzi-worker thread to send all data to loki. I tried to simulate a race condition by suspending the tjahzi-worker thread before the main process creates its last log message and calls then LogManager.shutdown(). After the shutdown the main process exits and I let the tjahzi-worker thread continue. It does still some work before terminating but the message does not arrive in loki.
Is this test valid? Or does the manual suspend/resume something different than a "normal" delay?
(Actually without suspending the thread, the message arrives at loki)

**Reply by tkowalcz** [reply]
Thanks for reporting this. I will definitely take a look but I am swamped right now. Hopefully next week I will let you know my findings.

**Reply by thgau** [reply]
No need to hurry. As I said: it usually works...

**Reply by tkowalcz** [reply]
Hi. I investigated more deeply how Log4j2 does the shutdown and here are my findings. See also [this](https://stackoverflow.com/questions/25877102/how-to-properly-shutdown-log4j2) (a little old) discussion.

In a "normal" application with `public static void main(String[] args)` I need to call `LogManager.shutdown()` to init a proper shutdown sequence. Then application exits cleanly.
When I change Tjahzi threads to be daemon threads then after `main` exits the shutdown hook registered by Log4j2 is executed and whole application exits cleanly. So this is definitely the way to go.

**Reply by tkowalcz** [reply]
As for your experiment. After log4j2 shutdown LokiAppender::stop method is called and Tjahzi attempts to shut down its threads. There is a `shutdownTimeoutSeconds` configuration value that controls how long it waits for log shipper thread to end. After that timeout the shutdown sequence continues. If you extend value of `shutdownTimeoutSeconds` to very large number will the message consistently appear in Loki?

**Reply by thgau** [reply]
Actually I can no more reproduce the problem. I do not know what was different a week ago. In the meantime I switched between the nodep and the plain version and from 0.9.17 to 0.9.18 but in none of these versions I get the problem any more. Currently it works fine, the last log message arrives and then the tjahzi-worker exits. I am sorry that you spent time in this problem.

**Reply by tkowalcz** [reply]
No problem. I learned few more things about log4j, it's all for the better.
